### PR TITLE
Use array pooling and various optimizations to reduce allocations

### PR DESCRIPTION
**Pull request by atravita-mods**
Basically, this leverages array pooling to reduce the amount of allocations.

I also switched array copying over to Array.Copy and adjusted how logging works a bit to reduce allocations there too.

There's more that could in theory be pooled (basically every file read/file write could be moved to using recyclable streams).

### CODE REVIEW

File: src/SMAPI/Framework/Content/AssetDataForImage.cs [hunk 1]
```diff
@@ -46,45 +48,71 @@ public void PatchImage(IRawTextureData source, Rectangle? sourceArea = null, Rec
...
-                    sourceData = new Color[areaWidth * areaHeight];
-                    int i = 0;
-                    for (int y = areaY, maxY = areaY + areaHeight - 1; y <= maxY; y++)
+                    int pixelCount = areaWidth * areaHeight;
+                    sourceData = ArrayPool<Color>.Shared.Rent(pixelCount);
+
+                    if (areaX == 0 && areaWidth == source.Width)
+                    {
+                        // shortcut copying because the area to copy is contiguous. This is
+                        // probably uncommon, but Array.Copy benefits a lot.
```
Comment (atravita-mods) [inline src/SMAPI/Framework/Content/AssetDataForImage.cs]: Occurs to me that there's really no point copying if the area is contiguous - I can rewrite this to pass in some sort of "start row" or "start index" instead.

Comment (atravita-mods) [reply src/SMAPI/Framework/Content/AssetDataForImage.cs]: and done.

**Comment by Pathoschild** [comment]
(I just rebased the branch to remove the merges.)

**Review by Pathoschild — APPROVED** [review]

**Comment by Pathoschild** [comment]
Thanks for the help! Merged into `develop` for the upcoming SMAPI 3.17.0.

**Pull request merged**